docs/leysin-2020-design-decisions.md

```diff
@@ -0,0 +1,295 @@
+
+```C
+// universal
+static inline HPy HPy_FromVoidP(HPyContext ctx, void *p) { return (HPy){(HPy_ssize_t)p}; }
+static inline void* HPy_AsVoidP(HPyContext ctx, HPy h) { return (void*)h._i; }
+
+// cpython (the -/+4 is to avoid people casting it to PyObject)
+static inline HPy HPy_FromVoidP(HPyContext ctx, void *p) { return (HPy){(HPy_ssize_t)p - 4}; }
+static inline void* HPy_AsVoidP(HPyContext ctx, HPy h) { return (void*)h._o + 4; }
+```
+
+/*
+ * Should we implement HPy_Dump?
+ *
+ * Question: What should it print and where should it print it?
+ *
+ * Decision: It's useful for debugging if all macros are also available as functions
+ *           definitions.
+ *
+ * Decision: It should dump to stderr just like PyObject_Dump.
+ */
+
+/*
+ * How do we silence warnings from using HPy_METH_KEYWORDS?
+ *
+ * Decision: Write a cast to HPyMeth.
+ */
+
+{"add_ints_kw", (HPyMeth) add_ints_kw, HPy_METH_KEYWORDS, ""}
+
+/*
+ * How should HPyErr_Format be implemented? Should we avoid va_args?
+ *
+ * Decision:
+ */
+
+HPyErr_Format(ctx, const char *fmt, ...) {
+  const char *msg = HPyStr_Format(ctx, fmt, ...);
+  HPyErr_SetString(msg);
+}
+
+ctx->ctx_HPyErr_Format(???)
+
+/*
+ * Should make specifying values for optional arguments & dup / closing them
+ * less messy?
+ *
+ * Decision: Right now, No. In the future, someone should invent
+ *           ARGUMENT_CLINIC for HPy.
+ */
+
+/*
+ * Should we rename "struct _object*" to "PyObject*" and "_HPy_PyCFunction"
+ * to "PyCFunction" in universal/hpy.h?
+ *
+ * Decision: No. This would generate a warning if one imports Python.h.
+ *
+ * Decision: Comment in the code that this _object* is PyObject* and why we
+ *           cannot call it that.
+ *
+ * Decision: In user documentation, just call ing PyObject *.
+ */
+
+/*
+ * How should the API pass around exceptions?
+ *
+ * Decision: Follow CPython for each API call.
+ *
+ */
+
+// E.g. API call returns HPy:
+
+HPy h = HPyLong_FromLong(ctx, 5);
+if (HPy_IsNull(ctx, h)) {
+  // handle error
+  // python error has been set by the API call
+}
+
+// or
+
+HPy h = HPyLong_FromLong(ctx, 5);
+if (HPyErr_Occurred(ctx)) {
+  // handle error
+  // python error has been set by the API call
+}
+
+// E.g. API call returns a value that is not an HPy:
+
+long l = HPyLong_AsLong(ctx, h);
+if (l == -1 && HPyErr_Occurred(ctx)) {
+  // handle error
+  // python error has been set by the API call
+}
+
+// E.g. API call returns a success or error flag:
+
+// int error = HPyArg_Parse(...);
+if (!HPyArg_Parse(...)) {
+  // handle error
+  // python error has been set by the API call
+}
+
+/*
+ * How should support for creating custom Python types look?
+ *
+ * Decisions:
+ *
+ *
+ */
+
+// When using C-API:
+
+typedef struct {
+  PyObject_HEAD
+     /* Type-specific fields go here. */
+     PyObject *x;
+     PyObject *y;
+     double *data;
+  int size;
+} PointObject;
+
+// When using HPy:
+
+typedef struct {
+    HPyObject_HEAD
+    double x;
+    double y;
+} HPy_Point;
```

Comment:
Just a remark: it seems to me that it's confusing that consumers of the HPy API define variables starting with HPy. Could names starting with HPy be reserved for the HPy API? The equivalent of `PointObject` could be `PointHPy`. At least, it's clear that `PointHPy` is not part of the HPy API but a user defined struct. I don't know if it makes sense?

Reply:
I did a quick search of `PyType_Ready` on github to check how people are currently calling their C types. A small number of extensions use the `Py` prefix, but the majority seems not to use any particular prefix, so +1 to remove the `HPy_` prefix.

Reply:
Agreed.